### PR TITLE
6428 set canmount=off on unmounted filesystem tries to unmount children

### PR DESCRIPTION
**Pull request by avg-I**
This change is in FreeBSD already:
- https://svnweb.freebsd.org/base?view=revision&revision=297521
- https://svnweb.freebsd.org/base?view=revision&revision=304520

### CODE REVIEW

**Comment by prakashsurya** [comment]
@zettabot go

**Comment by prakashsurya** [comment]
This failed 3 of the zfstest suite. Specifically, these 3 tests:
```
$ cat rsend/rsend_021_pos/merged
ASSERTION: Verify resumability of a full and incremental ZFS send/receive  with the -e (embedded) flag
SUCCESS: /usr/sbin/zfs destroy -r testpool.101281/sendfs
cannot mount '/testpool.101281/sendfs': directory is not empty
filesystem successfully created, but not mounted
SUCCESS: /usr/sbin/zfs destroy -r testpool.101281/stream
SUCCESS: /usr/sbin/zfs create -o compress=lz4 testpool.101281/stream
WARNING: could not send testpool.101281/sendfs@a: does not exist total estimated size is 0
ERROR: eval /usr/sbin/zfs send -v -e testpool.101281/sendfs@a >/testpool.101281/stream/1 exited 1
NOTE: Performing local cleanup via log_onexit (cleanup_pool testpool1.101281)
SUCCESS: /usr/bin/rm -rf /backdir-rsend/*
SUCCESS: /usr/sbin/zfs destroy -Rf testpool1.101281
SUCCESS: /usr/bin/rm -rf /testpool1.101281/*

$ cat rsend/rsend_022_pos/merged
ASSERTION: Verify resumability of an incremental ZFS send/receive with ZFS  bookmarks
SUCCESS: /usr/sbin/zfs destroy -r testpool.101281/sendfs
cannot mount '/testpool.101281/sendfs': directory is not empty
filesystem successfully created, but not mounted
SUCCESS: /usr/sbin/zfs destroy -r testpool.101281/stream
SUCCESS: /usr/sbin/zfs create -o compress=lz4 testpool.101281/stream
cannot open 'testpool.101281/sendfs@a': dataset does not exist usage: bookmark <snapshot> <bookmark> For the property list, run: zfs set|get For the delegated permission list, run: zfs allow|unallow
ERROR: /usr/sbin/zfs bookmark testpool.101281/sendfs@a testpool.101281/sendfs#bm_a exited 2
NOTE: Performing local cleanup via log_onexit (cleanup_pool testpool1.101281)
SUCCESS: /usr/bin/rm -rf /backdir-rsend/*
SUCCESS: /usr/sbin/zfs destroy -Rf testpool1.101281
SUCCESS: /usr/bin/rm -rf /testpool1.101281/*

$ cat rsend/rsend_024_pos/merged
ASSERTION: Verify resumability of a full ZFS send/receive with the source  filesystem unmounted
SUCCESS: /usr/sbin/zfs destroy -r testpool.101281/sendfs
cannot mount '/testpool.101281/sendfs': directory is not empty
filesystem successfully created, but not mounted
SUCCESS: /usr/sbin/zfs destroy -r testpool.101281/stream
SUCCESS: /usr/sbin/zfs create -o compress=lz4 testpool.101281/stream
cannot unmount 'testpool.101281/sendfs': not currently mounted
ERROR: /usr/sbin/zfs unmount testpool.101281/sendfs exited 1
NOTE: Performing local cleanup via log_onexit (cleanup_pool testpool1.101281)
SUCCESS: /usr/bin/rm -rf /backdir-rsend/*
SUCCESS: /usr/sbin/zfs destroy -Rf testpool1.101281
SUCCESS: /usr/bin/rm -rf /testpool1.101281/*
```

**Comment by ahrens** [comment]
I agree with the bug report (that it is a problem).  But this change seems to be changing behavior in other ways too.  Specifically, `zfs set canmount=on` should *not* attempt to mount anything.  I would guess that's what is causing the test failures.

File: usr/src/lib/libzfs/common/libzfs_dataset.c [hunk 1]
```diff
@@ -1594,11 +1594,17 @@ zfs_prop_set_list(zfs_handle_t *zhp, nvlist_t *props)
 		assert(cl_idx < nvl_len);
 		/*
 		 * We don't want to unmount & remount the dataset when changing
-		 * its canmount property to 'on' or 'noauto'.  We only use
-		 * the changelist logic to unmount when setting canmount=off.
+		 * its canmount property.  We only use the changelist logic to
+		 * unmount when setting canmount=off for a mounted filesystem
+		 * or when setting canmount=on for an unmounted filesystem.
```
Comment (ahrens) [inline usr/src/lib/libzfs/common/libzfs_dataset.c]: `zfs set canmount=on` should not attempt to mount anything.

**Comment by avg-I** [comment]
@ahrens, hrm, I am not sure where I got the idea that canmount=on should mount a dataset...
Sigh, as you can my FreeBSD change consists of two commits where the second introduced that behavior.

**Comment by ahrens** [comment]
@zettabot go

**Comment by avg-I** [comment]
@ahrens does the latest code look better?

**Comment by avg-I** [comment]
@ahrens , could you please revisit this PR (the 'question' tag is from you)?  Thank you!

**Pull request closed without merging**